DSpace/DSpace
# Support virtual metadata for authority-controlled values

## References
* Includes the (as yet unmerged) changes in https://github.com/DSpace/DSpace/pull/9845 (a small PR that just enhances the core AuthorityValue class to support additional metadata maps, like its subclasses)
 (I may just close that one in favour of including all changes here)
* Includes the (as yet unmerged) changes in https://github.com/DSpace/DSpace/pull/9842 (a small PR that removes unnecessary context parameter from AuthorityValueService find methods)

## Work in progress
* Integration / Unit tests are still in development

## Description
Changes and new simple beans to support virtual metadata for authority-controlled values.

If enabled, and if `enableVirtualMetadata` is true, the `ItemServiceImpl#getMetadata(..., enableVirtualMetadata)` method will additionally populate the metadata list with values collected from linked authority records.

The Spring configuration and beans are set up in a very similar way to the relationship-based classes, with one key difference being the lack of ability to set a designated `place` for the virtual metadata value (instead, the place is calculated at the time of population to ensure all virtual values are effectively appended).

The result of this is that  for an authority-controlled value like:

`dc.subject = Twain, Mark [authority = '12345']`

The authority key will be resolved to a record, the additional metadata in the authority record retrieved and mapped to virtual field names, and populated with the list of returned metadata.

## Dev notes

* As noted, place has to be calculated and not specified, as there is no equivalent of the places in related item metadata values to use as a basis
* The `MetadataValue#getID()` method must return an integer. Authority record field values cannot be addressed this way. Nor can relationship-based virtual metadata as they currently return the relationship ID, not a value ID (and they might be a concatenation or derived value anyway, not resolvable to a single metadata value).
To avoid any accidental mix-up with a real metadata value that is *not* this virtual authority metadata value, `-1` is always returned from the new `VirtualMetadataValue` subclass of `MetadataValue`.

## Instructions for Reviewers

List of changes in this PR:
* Changes to `AuthorityValue` to support maps of additional metadata (map of of lists of strings) like `PersonAuthorityValue`, `ORCIDv3AuthorityValue`, etc., just to bring the super class in line with implementations and make it more flexible
* New `VirtualMetadataValue` interface for type checking and inheritance (with both `RelationshipMetadataValue` and `AuthorityVirtualMetadataValue` inheriting), and a default safe `getID()` override
* New set of classes and config to implement configuration, collection, population of authority virtual metadata in a similar way to relationship virtual metadata:
   * `AuthorityVirtualMetadataValue`
   * `AuthorityVirtualMetadataConfiguration`
   * `AuthorityCollected`
   * `AuthorityVirtualMetadataPopulator`
   * Additional spring configuration included as example in `virtual-metadata.xml`
* Modify `ItemServiceImpl#getMetadata` to populate authority virtual metadata after relationship virtual metadata, if enabled
* Integration and unit test coverage (**TODO**)

To test:
* Set up authority control where the records will store some more complex data than a simple string value (e.g. ORCID author authority)
* Add some authors
* Set up virtual metadata field mapping as appropriate in `virtual-metadata.xml`
    * For example, include the orcid ID value, or an institutional affiliation from the Solr authority record 
* After building and deploying, visit the item page and view the full metadata page to check if the virtual metadata is present

## Checklist
